jq255.md

```diff
@@ -0,0 +1,1405 @@
+     eg = -3
+     ug = -1
+```
+
+For jq255s, the curve parameters are as follows:
+
+```
+jq255e:
+
+  q = 2^255 - 3957
+  a = -1
+  b = 1/2 = 2^254 - 1978
+  a' = -2*a      = 2
+  b' = a^2 - 4*b = -1
+  equation: e^2 = -u^4 + 2*u^2 + 1
+  curve order: n = 2*r
+     r = 2^254 + 56904135270672826811114353017034461895
+  conventional generator: G = (eg, ug)
+     eg = 6929650852805837546485348833751579670837850621479164143703164723313568683024
+     ug = 3
+```
+
+## Group Element Operations
+
+In this section, practical procedures for computing point operations are
+presented using pseudocode in a syntax compatible with the Python language
+and the [SageMath] mathematics system. This syntax closely follows the
+abstract mathematical notation otherwise used in this document, with the
+following exceptions:
+
+  - Exponentiation uses the `**` operator (instead of `^`).
+  - Points with extended coordinates `(E:Z:U:T)` become
+    tuples `(E,Z,U,T)`.
+
+The pseudocode assumes that the following functions are available:
+
+  - `GFq(j)` converts the integer `j` to a field element by reducing
+    `j` modulo `q`.
+
+  - `int(x)` converts the field element `x` into an integer by using its
+    unique representant in the `0` to `q-1` range.
+
+  - `half(x)` divides the field element `x` by two. It is usually
+    implemented by a right-shift by 1 bit, then the conditional addition
+    of the constant `(q+1)/2` if the dropped bit was non-zero. It is
+    considered to be a "fast" operation (i.e. at least as fast as an
+    addition in the field).
+
+  - `invert(x)` returns the multiplicative inverse of `x` in the field.
+    Zero has no multiplicative inverse, but it is sometimes convenient
+    for constant-time implementations to return `0` as a (formal) inverse
+    of `0`; this is obtained naturally if inversion is implemented using
+    Fermat's Little Theorem (i.e. raising `x` to the power `q-2`). This
+    feature is not needed for the formulas described in this document.
+
+  - `sgn(x)` returns the "sign" of `x`, as defined previously: the
+    sign is the value of the least significant bit of `int(x)`.
+
+  - `sqrt(x)` returns either a square root of `x` (if `x` is indeed
+    a square in the finite field), or `False` (if `x` is not a square
+    in the finite field). **Important:** when `x` is a square, the
+    returned square root is non-negative.
+
+  - `is_square(x)` return `True` if `x` is a square in the finite field,
+    or `False` otherwise.
+
+  - `scalar_to_int(s)` and `int_to_scalar(j)` compute the same
+    operations as `int()` and `GFq()`, respectively, but for scalars,
+    which are integers modulo the prime `r`.
+
+  - `secure_random(m)`, for an integer `m >= 1`, returns a sequence
+    of `m` random bytes obtained from a cryptographically secure source,
+    with a uniform probability distribution. This is used only for
+    private key generation in the specification of high-level protocols.
```

Comment:
Does this need more detail, hm

Reply:
You mean adding a link to RFC 4086? I am not entirely sure this is a good idea. I am tempted to say that if an implementer generates private keys, sees the words "cryptographically secure source", and goes for Mersenne Twister, then that calls for some educational session that will exceed the scope of the spec...

Reply:
Not necessarily a link to RFC 4086, perhaps an example ("such as sourcing from /dev/urandom") may suffice

Reply:
Seems fine. Done.